paddle/fluid/framework/program_processing.cc

```diff
@@ -0,0 +1,135 @@
+  // NOTE: Here assumes that all variables are input or output of Ops,
+
+  std::set<std::string> removed_inner_inputs;
+
+  for (OpDesc *op : current_block.AllOps()) {
+    for (auto iname : op->InputNames()) {
+      for (auto in_var_name : op->Input(iname)) {
+        VLOG(3) << "in_var_name:" << in_var_name;
+        if (inner_outputs->find(in_var_name) == inner_outputs->end()) {
+          VLOG(3) << "insert inner_inputs_name:" << in_var_name;
+          inner_inputs->insert(in_var_name);
+        }
+      }
+    }
+
+    for (auto oname : op->OutputNames()) {
+      for (auto out_var_name : op->Output(oname)) {
+        VLOG(3) << "insert out_var_name:" << out_var_name;
+        inner_outputs->insert(out_var_name);
+      }
+    }
+  }
+
+  // Step2: Remove variable created in current control flow block.
+  BlockDesc *parent_block = current_block.ParentBlock();
+  VarDesc *current_block_var;
+
+  if (parent_block) {
+    for (auto in_var_name : *inner_inputs) {
+      VLOG(3) << "recursively find var:" << in_var_name;
+      VarDesc *parent_block_var = parent_block->FindVarRecursive(in_var_name);
+      if (current_block.HasVar(in_var_name))
+        current_block_var = current_block.FindVar(in_var_name);
+      if (parent_block_var == nullptr && current_block_var &&
+          current_block_var->GetType() == proto::VarType::LOD_TENSOR) {
+        VLOG(3) << "remove inner var:" << in_var_name;
+        removed_inner_inputs.insert(in_var_name);
+      }
+    }
+    std::set<std::string> inner_inputs_;
+    std::set_difference(inner_inputs->begin(), inner_inputs->end(),
+                        removed_inner_inputs.begin(),
+                        removed_inner_inputs.end(),
+                        inserter(inner_inputs_, inner_inputs_.begin()));
+    inner_inputs->swap(inner_inputs_);
+  }
+}
+
+void ProgramProcessor::AddDepToBlockOp(const BlockDesc &block) {
+  VLOG(3) << "Op size:" << block.AllOps().size();
+  for (OpDesc *op : block.AllOps()) {
+    if (op->HasAttr("sub_block")) {
+      auto op_type = op->Type();
+      BlockDesc *sub_block = boost::get<BlockDesc *>(op->GetAttr("sub_block"));
+
+      // recursively processing
+      AddDepToBlockOp(*sub_block);
+
+      std::set<std::string> sub_inputs;
+      std::set<std::string> sub_outputs;
+      ProgramProcessor::GetInputsOutputsInBlock(*sub_block, &sub_inputs,
+                                                &sub_outputs);
+      VLOG(3) << "sub_inputs.size:" << sub_inputs.size();
+      VLOG(3) << "sub_outputs.size:" << sub_outputs.size();
+
+      auto *op_inputs = op->MutableInputs();
+      std::vector<std::string> *op_input_var_vec;
+      VLOG(3) << "op_type:>>>>>>" << op_type;
+      if (op_type.compare("while") == 0)
+        op_input_var_vec = &((*op_inputs)["kX"]);
+      else if (op_type.compare("conditional_block") == 0)
+        op_input_var_vec = &((*op_inputs)["kInputs"]);
+      else
+        // Only support while_op and conditinal_block_op now
```

Comment:
Please log a warning for this case, if our program runs here, it means something is wrong.

Reply:
Done.

Reply:
1. You shouldn't `continue` after `throw`
2. I suggest to log warning instead of throwing an exception in case of other operators may contain "sub_block" attributes.